[WIP] Multiperiod SWRO flowsheet

arao53 opened the pull request:
## Fixes/Resolves:

## Summary/Motivation:
A code example of a similar approach by [DISPATCHES](https://github.com/gmlc-dispatches/dispatches/blob/e5dbfbfbe4e0dd63881c8195af9e489b5ba8ad27/dispatches/models/fossil_case/ultra_supercritical_plant/storage/pricetaker_with_multiperiod_integrated_storage_usc.py)

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

arao53 committed:
Created additional variables and constraints for variable efficiency pump class

arao53 committed:
Added isothermal constraint and expr_if expression with pump flow-efficiency correlation.

arao53 committed:
Added tests

arao53 committed:
Scaling factors to unit model and test

arao53 committed:
Head ratio constraint and solution tests

arao53 committed:
Changed variable efficiency to be implemented as a config option of the main WaterTAP pump

arao53 committed:
Removed duplicates and added additional tests for coverage

arao53 committed:
Addressed feedback and altered tests

arao53 committed:
Revert "Addressed feedback and altered tests"

This reverts commit 4518e1831f1703d2de13d61944b0b97eabb9d38c.

arao53 committed:
Merge branch 'main' of https://github.com/arao53/watertap into variable-idaes-pump

arao53 committed:
Added configuration options for variable efficiency to the RO_with_energy_recovery flowsheet

arao53 committed:
Reconciled tests between none and flow type VariableEfficiency options

arao53 committed:
Created multiperiod model flowsheet base

arao53 committed:
Created multiperiod model with linking/ramp constraints

adam-a-a commented on watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:
```diff
@@ -45,32 +45,43 @@
+class VariableEfficiency(StrEnum):
+    none = "none"
+    flow = "flow"
+    flow_head = "flow_head"
```
Couldn't we import this from the modified pump model?

adam-a-a commented on watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:
```diff
@@ -82,9 +93,27 @@ def build():
+            }
+        )
+    else:
+        raise NotImplementedError("Variable efficiency type not recognized")
```
```suggestion
  
       
      m.fs.P1 = Pump(
          default={
              "property_package": m.fs.properties,
              "variable_efficiency": variable_efficiency,
          }
      )
      m.fs.P2 = Pump(
          default={
              "property_package": m.fs.properties,
              "variable_efficiency": variable_efficiency,
          }
      )
```
I think you can just pass the variable_efficiency arg into the pump config option, and error handling should/could be caught by the pump unit itself.

adam-a-a commented on watertap/examples/flowsheets/RO_with_energy_recovery/multiperiod_RO.py:
```diff
@@ -0,0 +1,181 @@
+    m = ConcreteModel()
+    m.ro_mp = swro.main(swro.VariableEfficiency.flow)
+
+    print("\nUnfixing design variables and setting bounds...")
```
Unfixing or fixing?

adam-a-a reviewed (commented):
Just a quick initial scan- I still have to go through the rest, but looks good so far!

arao53 committed:
Addressed feedback about variable efficiency case import and use

arao53 committed:
Refactored multiperiod work to its own directory under flowsheets

arao53 committed:
Import LMP data

arao53 committed:
Rearranged imports

arao53 committed:
Saving progress

arao53 committed:
Added solve, display, and compare to baseline for pricetaker_multiperiod_RO.py

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

arao53 committed:
Added check for existing enthalpy balance constraint and removed extra comments

arao53 committed:
Added tests. Variable type repeats isothermal pump tests and checks flow-efficiency correlation x4

arao53 committed:
Merge branch 'main' into variable-idaes-pump

arao53 committed:
Removed unfix for pump_efficiency

arao53 committed:
Merge branch 'variable-idaes-pump' of https://github.com/arao53/watertap into variable-idaes-pump

arao53 committed:
Resolved test error

arao53 committed:
Resolved issues with efficiency correlation by defining flow_ratio based on flow_vol_phase["Liq"] instead of flow_vol

lbianchi-lbl commented:
@arao53 @adam-a-a To test this PR with IDAES as if IDAES/idaes-pse#908 was already merged, you can set the `idaes-pse` requirement in `setup.py` on line 46 to `idaes-pse @ https://github.com/IDAES/idaes-pse/archive/refs/pull/908/merge.zip`.

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

arao53 commented:
@adam-a-a @lbianchi-lbl This flowsheet works as expected with the tag for [IDAES/idaes-pse#908](https://github.com/IDAES/idaes-pse/pull/908). This was done with commit d89fe42327687fda3542a6ba77d7eb99ef91c058.

arao53 committed:
Removed additional print statements and extra solve.

arao53 committed:
Tests for multiperiod_RO.py

arao53 committed:
Tests for pricetaker RO flowsheet

adam-a-a committed:
Merge branch 'main' into variable-idaes-pump

arao53 committed:
Merge branch 'variable-idaes-pump' of https://github.com/arao53/watertap into variable-RO-flowsheet

arao53 committed:
Restructured code and added tests

arao53 committed:
Test coverage and formatting

arao53 committed:
Merge branch 'watertap-org:main' into variable-RO-flowsheet

arao53 committed:
Merge branch 'main' of https://github.com/watertap-org/watertap into variable-RO-flowsheet

arao53 committed:
Merge branch 'variable-RO-flowsheet' of https://github.com/arao53/watertap into variable-RO-flowsheet

arao53 committed:
Fixed filepath in tests

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

arao53 committed:
Merge branch 'variable-RO-flowsheet' of https://github.com/arao53/watertap into variable-RO-flowsheet

arao53 committed:
Used os to find the path of the datafile to resolve test errors

arao53 committed:
Add carbon emissions and cost parameters to pricetaker scenario

arao53 committed:
Reformulating constraints

arao53 committed:
Applied a weighted avg. permeate tds constraint and switched to using get_solver() method for optimizer.

arao53 committed:
Additions to multiperiod flowsheet initialization

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

arao53 committed:
Reformatted with black and resolved a syntax error

arao53 committed:
Resolving additional merge conflicts between variable efficiency and configuration type options

arao53 committed:
Resolving discrepencies with intermediate tests.

arao53 committed:
Resolved merge conflict errors with tests

arao53 committed:
Saving progress. Tinkering with bounds and constraints.

arao53 committed:
Merge branch 'main' of https://github.com/watertap-org/watertap into variable-ro-flowsheet

arao53 committed:
Revised plotting for turbine configuration

arao53 committed:
Price multiplier parameter

arao53 committed:
Reformulated problem to fix daily capacity and make the objective linear with respect to design variables

bknueven commented on watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py:
```diff
@@ -0,0 +1,431 @@
+        expr=abs(sum([blk.water_prod for blk in t_blocks]) / daily_water_production - 1)
+        <= 1e-5,
+        doc="The water production over the day is fixed within a tolerance",
+    )
```
```suggestion
    m.permeate_yield = Constraint(
        expr=(daily_water_production -1e-5,
            sum(blk.water_prod for blk in t_blocks),
            daily_water_production + 1e-5),
        doc="The water production over the day is fixed within a tolerance",
    )
```
I'm surprised Pyomo ingests the `abs`, but in any case this is a nice linear constraint so it isn't necessary.

bknueven commented on watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py:
```diff
@@ -0,0 +1,431 @@
+    m.permeate_quality_lb = Constraint(
+        expr=(sum([blk.weighted_permeate_quality for blk in t_blocks]) >= 0),
+        doc="Flow-averaged permeate quality must be greater than 0 ppm",
+    )
```
Any reason not to enforce these block-by-block as opposed to at the aggregate level?

I guess if it goes into one large tank it's okay, but for Ipopt these constraints would be less troublesome enforced at every time step, I think.

arao53 committed:
Adjust objective function to use a fixed base capital cost for each timestep

arao53 committed:
Added permeate flowrate output in save_results

arao53 committed:
Add upper bound for capacity constraint.

arao53 committed:
Included params for demand surcharge

arao53 committed:
Incorporating demand and time based surcharges for electricity at the multiperiod level

arao53 committed:
Removed "demand charge" formulation and redundant comments

arao53 committed:
Adds comments and more explicit options for activiating and deactivating MP constraints.

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

arao53 committed:
Added an unfix dof function to building the multiperiod model.

arao53 committed:
Reverted example to use constant efficiency pump.

arao53 committed:
Removed process model initialization from initial multiperiod build.
MP initialization will be deferred to a custom initialization routine
in future commits.

arao53 committed:
Renamed blk to b for consistency across functions.

arao53 committed:
1. Added ramping parameters in the multiperiod model class to individual block.
2. Removed the dagget_CA_LMP_hourly_2015.csv file from the repository.
3. Changed RO flowsheet example to remove redundancy in setting the objective.

arao53 committed:
Removed redundant initializations and demand charge calculations.
Added dynamic parameters to new blocks.

arao53 committed:
Delete redundant initialization function.

arao53 committed:
Removed redundancies in the save_results function

arao53 committed:
Unfixed recovery ratio and defined a daily production requirement input.

arao53 committed:
Refining parameter definitions and save results.

arao53 committed:
Fixed electricity price per block and fixed flow volume in unfix dof function.

arao53 committed:
Merge branch 'main' into variable-RO-flowsheet

adam-a-a commented on watertap/examples/flowsheets/RO_multiperiod_model/multiperiod_RO.py:
```diff
@@ -0,0 +1,152 @@
+    m.fs.dynamic.ramping_rate = Param(initialize=0.7e5,
+                                        units= pyunits.Pa/pyunits.s,
+                                        mutable=True,
+                                        doc="Slowest rate at which pressure can be ramped up or down")
```
```suggestion
                                        doc="Fastest rate at which pressure can be ramped up or down")
```

adam-a-a committed:
Merge branch 'main' into variable-RO-flowsheet

adam-a-a committed:
Merge branch 'main' into variable-RO-flowsheet

adam-a-a commented:
Closing this PR as it is outdated, and we will revamp this later in a new PR to provide core functionality for flexible desal.
